compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIELogicalObjFifoSplittingUtils.cpp

```diff
@@ -0,0 +1,332 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "AMDAIELogicalObjFifoSplittingUtils.h"
+
+#include "llvm/ADT/DenseMap.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/IR/Operation.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/Iterators.h"
+
+namespace mlir::iree_compiler::AMDAIE {
+
+/// Utility to verify that the split dimensions for L2 are contiguous.
+static LogicalResult verifySplitDimensionConstraint(
+    SmallVector<unsigned> &splitDimsSetForL2) {
+  unsigned dim = 0;
+  for (unsigned splitDim : splitDimsSetForL2) {
+    if (splitDim != dim) return failure();
+    ++dim;
+  }
+  return success();
+}
+
+/*
+  For L3 -> L2 DmaCpyNd :-
+  From offset (0,0) we are extracting one 4x4 memref.
+                  _______
+                |. . . .|
+                |. . . .|
+                |. . . .|
+                |. . . .|
+                ---------
+
+  After split we will extract four 2x2 memrefs.
+  So, the corresponding offsets will be :-
+  1. Offset (0,0) - extract 2x2 memref
+        ___
+      |. .|. .
+      |. .|. .
+      -----
+        . . . .
+        . . . .
+  2. Offset (0,2) - extract 2x2 memref
+            ___
+        . .|. .|
+        . .|. .|
+          -----
+        . . . .
+        . . . .
+  3. Offset (2,0) - extract 2x2 memref
+        . . . .
+        . . . .
+        ___
+      |. .|. .
+      |. .|. .
+      -----
+  4. Offset (2,2) - extract 2x2 memref
+        . . . .
+        . . . .
+            ___
+        . .|. .|
+        . .|. .|
+          -----
+
+  The following utility helps perform the computation of offsets for L3 source.
+*/
```

Comment:
```suggestion
/// This utility helps to perform the computation of offsets for L3 source.
///
/// Example:
/// For L3 -> L2 DmaCpyNd :-
/// From offset (0,0) we are extracting one 4x4 memref.
///                 _______
///               |. . . .|
///               |. . . .|
///               |. . . .|
///               |. . . .|
///               ---------
/// After split we will extract four 2x2 memrefs.
/// So, the corresponding offsets will be :-
/// 1. Offset (0,0) - extract 2x2 memref
///       ___
///     |. .|. .
///     |. .|. .
///     -----
///       . . . .
///       . . . .
/// 2. Offset (0,2) - extract 2x2 memref
///           ___
///       . .|. .|
///       . .|. .|
///         -----
///       . . . .
///       . . . .
/// 3. Offset (2,0) - extract 2x2 memref
///       . . . .
///       . . . .
///       ___
///     |. .|. .
///     |. .|. .
///     -----
/// 4. Offset (2,2) - extract 2x2 memref
///       . . . .
///       . . . .
///           ___
///       . .|. .|
///       . .|. .|
///         -----
```

Reply:
Also, doesn't this example belong to the documentation of the split function below?

Reply:
This is just about L3 -> L2 because for L2->L1 we do not modify non-split dimension offsets.